File: huipyo/src/week03/boj5107/Main.java
/**
 *  마니또
 *  https://www.acmicpc.net/problem/5107
 *  [ 효율성 ]
 *  - 메모리: 18552KB
 *  - 시간 : 240ms
 */

package week03.boj5107;

import java.util.*;
import java.util.function.Consumer;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n, no = 1;

        while ((n = scanner.nextInt()) != 0) {
            int count = 0;
            Map<String, List<String>> graph = initGraph(n, scanner);
            Set<String> visited = new HashSet<>();
            Consumer<String> bfs = start -> {
                Queue<String> queue = new ArrayDeque<>(List.of(start));

                while (!queue.isEmpty()) {
                    String current = queue.poll();

                    for (String next : graph.get(current)) {
                        if (visited.contains(next)) continue;
                        visited.add(next);
                        queue.offer(next);
                    }
                }
            };

            for (String s : graph.keySet()) {
                if (!visited.contains(s)) {
                    bfs.accept(s);
                    count++;
                }
            }
            System.out.println(no++ + " " + count);
        }
        scanner.close();
    }

    private static Map<String, List<String>> initGraph(int n, Scanner scanner) {
        return new HashMap<>() {
            {
                for (int i = 0; i < n; i++) {
                    String first = scanner.next(), second = scanner.next();

                    computeIfAbsent(first, s -> new ArrayList<>()).add(second);
                    computeIfAbsent(second, s -> new ArrayList<>()).add(first);
                }
            }
        };
    }
}

File: huipyo/src/week03/boj5107/Main.js
/**
 *  마니또 (페어프로그래밍 효영/희표)
 *  https://www.acmicpc.net/problem/5107
 *  [ 효율성 ]
 *  - 메모리: 9412KB
 *  - 시간 : 116ms
 */

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
let input = fs.readFileSync(filePath).toString().trim().split('\n');
let num = 0;

while (input.length) {
    let graph = {};
    let nC = Number(input.shift());

    if (nC === 0) break;
    num++;
    for (let i = 0; i < nC; i++) {
        let [first, second] = input.shift().split(' ');

        if (!graph[first]) graph[first] = [];
        if (!graph[second]) graph[second] = [];
        graph[first].push(second);
        graph[second].push(first);
    }
    let count = 0;

    // 방문 여부를 나타내는 Set 초기화
    let visited = new Set();

    // BFS 알고리즘 구현
    function bfs(start) {
        let queue = [start];

        while (queue.length > 0) {
            let current = queue.shift();

            for (let next of graph[current]) {
                if (visited.has(next)) continue;
                visited.add(next);
                queue.push(next);
            }
        }
    }

    for (let node in graph) {
        if (!visited.has(node)) {
            bfs(node);
            count++;
        }
    }
    console.log(`${num} ${count}`);
}